### ISSUE TITLE
Accessing per-sample loss during training

### ISSUE THREAD

**Issue by yoavhalevi** [ultralytics/ultralytics#20329]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hey,

I am designing a sampler that does hard samples mining, meaning that each epoch, a percentage of the hardest samples from the previous epoch will be upsampled in the new epoch.
The 'hard' definition is based on the loss and for that I need the per-sample los.

Currently, for YOLO 11, the `BaseModel.loss()` lazily instantiates the loss via `init_criterion()`, which typically returns a `v8DetectionLoss` instance. The loss components (box, cls, DFL) are summed over the batch internally before being returned.

**My question:**  
*What is the recommended or cleanest way to access per-sample loss values, ideally without modifying the core source code?*

I've considered wrapping the loss object via a custom `init_criterion()` override in a model subclass — is that the best approach, or is there a built-in or supported way to achieve this?

Also, I’d like to avoid re-computing the loss in a callback, since that would introduce unnecessary overhead and slow down training.

Appreciate any insights or suggestions!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @yoavhalevi, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users, where you can find many helpful [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. Many common questions are already answered there.

This is an automated response to help you get started. An Ultralytics engineer will review your question and assist you further soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much detail as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best! For real-time chat, join our [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Hi @yoavhalevi,

You're right about the current loss aggregation in YOLOv11. The best approach would be to subclass the loss function rather than modifying core code. Here's what I recommend:

Create a custom loss class that extends `v8DetectionLoss` and overrides the `__call__` method to capture per-sample losses before summation. You can then use this with a custom model that overrides `init_criterion()`.

The key modification would be in the loss class, focusing on lines like `loss[0] *= self.hyp.box` and `return loss * batch_size, loss.detach()` to store individual sample losses before multiplication and summation.

This approach maintains compatibility with the training pipeline while giving you access to the per-sample metrics you need for hard sample mining.
